packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts

```diff
@@ -58,55 +70,79 @@ function sortByInputIndexThenByName(
  *
  * Since `source1` has no run data and no pinned data it's skipped in favor of
  * `source2` for the for input.
+ * It will become it's own group that is marked as `complete: false`
  *
- * So this will return 1 group:
- * 1. source2 and source3
+ * So this will return 2 group:
+ * 1. source2 and source3, `complete: true`
+ * 2. source1, `complete: false`
  */
 export function getSourceDataGroups(
 	graph: DirectedGraph,
 	node: INode,
 	runData: IRunData,
 	pinnedData: IPinData,
-): GraphConnection[][] {
+): Group[] {
 	const connections = graph.getConnections({ to: node });
 
 	const sortedConnectionsWithData = [];
+	const sortedConnectionsWithoutData = [];
 
 	for (const connection of connections) {
 		const hasData = runData[connection.from.name] || pinnedData[connection.from.name];
 
 		if (hasData) {
 			sortedConnectionsWithData.push(connection);
+		} else {
+			sortedConnectionsWithoutData.push(connection);
 		}
 	}
 
 	sortedConnectionsWithData.sort(sortByInputIndexThenByName);
+	sortedConnectionsWithoutData.sort(sortByInputIndexThenByName);
 
-	const groups: GraphConnection[][] = [];
-	let currentGroup: GraphConnection[] = [];
+	const groups: Group[] = [];
+	let currentGroup: Group = newGroup();
 	let currentInputIndex = -1;
 
-	while (sortedConnectionsWithData.length > 0) {
-		const connectionWithDataIndex = sortedConnectionsWithData.findIndex(
-			// eslint-disable-next-line @typescript-eslint/no-loop-func
-			(c) => c.inputIndex > currentInputIndex,
-		);
-		const connection: GraphConnection | undefined =
-			sortedConnectionsWithData[connectionWithDataIndex];
-
-		if (connection === undefined) {
-			groups.push(currentGroup);
-			currentGroup = [];
-			currentInputIndex = -1;
-			continue;
+	while (sortedConnectionsWithData.length > 0 || sortedConnectionsWithoutData.length > 0) {
```

Comment:
Is there a risk that this might cause an infinite loop in some obscure scenario?

Reply:
Yeah, that's hard to follow. 🤔 

Technically it should always terminate unless it contains a connection with a negative `inputIndex`.

I could iterate over all connections, find the lowest and highest `inputIndex`es and use them inside a for loop 🤔

Reply:
I tried rewriting it. It just turns messier than this. Let me know if you have an idea, otherwise I think the tests are pretty complete now.